modules/031-linstor/images/linstor-pools-importer/main.go

```diff
@@ -0,0 +1,386 @@
+		if err != nil {
+			klog.Fatalln("Failed to get LINSTOR node", err)
+		}
+		_, err = lc.Nodes.GetStoragePool(ctx, nodeName, storagePool.StoragePoolName)
+		if err != nil && err == lclient.NotFoundError {
+			err = lc.Nodes.CreateStoragePool(ctx, nodeName, storagePool)
+			if err != nil {
+				sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeWarning, "Failed", "Failed to create LINSTOR storage pool"+err.Error())
+				klog.Fatalln("Failed to create LINSTOR storage pool", err)
+			}
+			sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeNormal, "Created", "Created LINSTOR storage pool: "+nodeName+"/"+storagePool.StoragePoolName)
+		} else if err != nil {
+			klog.Fatalln("Failed to get LINSTOR storage pool", err)
+		}
+
+		// Get the maximum number of available replicas
+		opts := lclient.ListOpts{StoragePool: []string{storagePool.StoragePoolName}}
+		sp, err := lc.Nodes.GetStoragePoolView(ctx, &opts)
+		if err != nil {
+			klog.Fatalln("Failed to list LINSTOR storage pools", err)
+		}
+		replicasNum := len(sp)
+		if replicasNum > maxReplicasNum {
+			replicasNum = maxReplicasNum
+		}
+
+		// Create StorageClasses in Kubernetes
+		for r := 1; r <= replicasNum; r++ {
+			storageClass := newKubernetesStorageClass(&storagePool, r)
+			err = kc.Get(ctx, types.NamespacedName{Name: storageClass.GetName()}, &storageClass)
+			if err != nil && errors.IsNotFound(err) {
+				err = kc.Create(ctx, &storageClass)
+				if err != nil {
+					sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeWarning, "Failed", "Failed to create Kubernetes storage class"+err.Error())
+					klog.Fatalln("Failed to create Kubernetes storageClass", err)
+				}
+				sendKubernetesEvent(ctx, kc, &owner, v1.EventTypeNormal, "Created", "Created Kubernetes storage class: "+storageClass.Name)
+			} else if err != nil {
+				klog.Fatalln("Failed to get Kubernetes storage class:", err)
+			}
+		}
+	}
+}
+
+// Sends event to Kubernetes
+func sendKubernetesEvent(ctx context.Context, kc kclient.Client, owner *v1.ObjectReference, eventType, reason, message string) {
+	event := newKubernetesEvent(owner, eventType, reason, message)
+	err := kc.Create(ctx, &event)
+	if err != nil {
+		klog.Fatalln("Failed to create event", err)
+	}
+}
+
+// Makes loop over storage pool candidates, retruns channel of changed ones
+func runCandidatesLoop(ctx context.Context, f func() ([]lclient.StoragePool, error), delay time.Duration) <-chan lclient.StoragePool {
+	ch := make(chan lclient.StoragePool)
+	go func() {
+		var oldCandidates []lclient.StoragePool
+		for {
+			select {
+			case <-ctx.Done():
+				close(ch)
+				return
+			default:
+				candidates, err := f()
+				if err != nil {
+					klog.Fatalln("failed to load candidates:", err)
+				}
+			LOOP:
+				for _, candidate := range candidates {
+					for _, oldCandidate := range oldCandidates {
+						if candidate.StoragePoolName == oldCandidate.StoragePoolName {
+							continue LOOP
+						}
+					}
+					ch <- candidate
+				}
+				oldCandidates = candidates
```

Comment:
But this will not trigger the reconfiguration in case if storagepool was already seen before and then recreated for some reason.

But maybe that's OK. I need to think about it.

Reply:
It could be a map with a deadline to review the pool again after some time, e.g. 5 min

```go
// assuming delay is 1s
timeout := 5*time.Minute
seen := make(map[string]time.Time)

for {
	now := time.Now()
	for _, cand := range getCandidates() {
		time.Sleep(delay)

		when, ok := seen[cand.StoragePoolName] 
		if ok && when.Before(now) {
			continue
		}
		seen[cand.StoragePoolName] = now.Add(timeout)

		ch <- cand
	}
}
```